crates/sui-json-rpc/src/coin_api.rs

```diff
@@ -38,130 +38,91 @@ impl CoinReadApi {
-    fn multi_get_coin(
-        &self,
-        coins: &[ObjectRef],
-        coin_type_tag: Option<&TypeTag>,
-    ) -> Result<Vec<Result<SuiCoin, Error>>, Error> {
-        let o = self
-            .state
-            .database
-            .multi_get_object_by_key(&coins.iter().map(ObjectKey::from).collect::<Vec<_>>())?;
-        // conversion from TypeTag to string is expensive, so do it outside the loop if we already know the coin type
-        // if coin_type_tag is None, we are getting a heterogenous mix of coins and we have no choice but to string-ify in the loop
-        let coin_type_str = coin_type_tag.map(|t| t.to_string());
-
-        Ok(o.into_iter()
-            .zip(coins)
-            .map(|(o, (id, version, digest))| {
-                let o = o.ok_or(UserInputError::ObjectNotFound {
-                    object_id: *id,
-                    version: Some(*version),
-                })?;
-
-                if let Some(move_object) = o.data.try_as_move() {
-                    let coin_type = coin_type_str.clone().unwrap_or_else(|| {
-                        o.type_()
-                            .unwrap()
-                            .type_params()
-                            .first()
-                            .unwrap()
-                            .to_string()
-                    });
-                    let balance = {
-                        let coin: Coin = bcs::from_bytes(move_object.contents())?;
-                        coin.balance.value()
-                    };
-
-                    Ok(SuiCoin {
-                        coin_type,
-                        coin_object_id: *id,
-                        version: *version,
-                        digest: *digest,
-                        balance,
-                        locked_until_epoch: None,
-                        previous_transaction: o.previous_transaction,
-                    })
-                } else {
-                    Err(Error::UnexpectedError(format!(
-                        "Provided object : [{}] is not a Move object.",
-                        o.id()
-                    )))
-                }
-            })
-            .collect())
-    }
-
-    async fn get_coins_internal(
+    fn get_coins_iterator(
         &self,
         owner: SuiAddress,
-        coin_type: Option<&TypeTag>,
-        // exclusive cursor if `Some`, otherwise start from the beginning
-        cursor: Option<ObjectID>,
+        cursor: (String, ObjectID),
         limit: Option<usize>,
-    ) -> Result<CoinPage, Error> {
-        // TODO: Add index to improve performance?
+        one_coin_type_only: bool,
+    ) -> anyhow::Result<CoinPage> {
         let limit = cap_page_limit(limit);
-        let mut coins = self
-            .get_owner_coin_iterator(owner, coin_type)?
-            .skip_while(|o| matches!(&cursor, Some(cursor) if cursor != &o.0))
-            // skip an extra b/c the cursor is exclusive
-            .skip(usize::from(cursor.is_some()))
```

Comment:
Could you confirm if we are still keeping the `exclusive cursor` semantics? meaning the query will start from the next item after the cursor

Reply:
yes, it's done here https://github.com/MystenLabs/sui/pull/10679/files/61d1006565866b7421b616aaf3ec56fb82652732#r1165847229